### ISSUE TITLE
flag to export all commands/flags/options as json for use with library bindings

### ISSUE THREAD

**Issue by lrstanley** [yt-dlp/yt-dlp#7958]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a feature unrelated to a specific site
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.07.06** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

I'm looking at potential options for creating a binding library for Go, though others may also be interested in this for other languages that don't easily have direct bindings to Python. Ideally, the library would cover:

* Auto-generated wrapper around an installed version of the yt-dlp cli (user provides yt-dlp, similar to how many projects wrap ffmpeg)
   * Potential version constraint checking to ensure drift isn't an issue
* Maps groups -> types, flags -> struct fields with proper type mappings
* Simplifies the resulting output (where possible) (could see https://github.com/yt-dlp/yt-dlp/issues/1659 also being super helpful, or any other kind of structured output)

The ideal scenario is to simply get access to the optparse info after flag parsing. I've setup some code with a local fork of yt-dlp, which extracts the option data as JSON from yt-dlp, with a hidden `--export-options` flag. That JSON data can then be used to generate a binding library that handles all of the above scenarios automatically and can also easily be kept updated/in parity.

An example of what this JSON would output: https://cdn.liam.sh/share/2023/08/yt-dlp-example-export.json

Can submit a PR if y'all think this is a useful enough feature -- would definitely help with using yt-dlp in a more programmatic way.

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
n/a
```

**Comment by pukkandan** [comment]
I have never seen any program provide such an output. Do you know of any?

**Comment by lrstanley** [comment]
I'm not aware of any either off the top of my head, as generating libraries isn't super common unless it's related to things like API clients/actual bindings (e.g. wrapping a C library, in Python or Go).

Most cli-wrapper libraries are often hand written because the wrapped cli tool has few flags. In **yt-dlp**'s case, it looks like it's just over 300 at the moment (and I only suspect it will grow in the future). This compared to something like ffmpeg, which is very commonly cli-wrapped, that has ~65 flags (excluding those relating to help output).

I can't think of another simple way of handling this, especially in the case of languages which the binding to Python isn't as good or non-existant. This approach feels very lightweight (~60 LOC) -- only needs to consume optparse, doesn't need to be maintained with each new flag added, etc.

I could patch yt-dlp during code generation (and hope that the patch logic doesn't break often), if it doesn't make sense to have this inside yt-dlp itself.

**Comment by lrstanley** [comment]
Starting testing things out and generating code based on the above json.

Repo: https://github.com/lrstanley/go-ytdlp

Example of what the client-side code would look like:

```go
package main

import (
	"context"
	"fmt"

	"github.com/lrstanley/go-ytdlp"
)

func main() {
	dl := ytdlp.New().
		Verbose().
		NoProgress().
		NoCallHome().
		NoCheckCertificates().
		RejectTitle(".*preview/.*").
		FormatSort("res,ext:mp4:m4a").
		RecodeVideo("mp4").
		NoPlaylist().
		NoOverwrites().
		Continue().
		Output("%(extractor)s - %(title)s.%(ext)s")

	res, err := dl.Run(context.TODO(), "https://www.youtube.com/watch?v=dQw4w9WgXcQ")
	if err != nil {
		panic(err)
	}

	fmt.Printf("%#v\n", res)
}
```

Example of some of the generated code (nothing crazy, ofc): https://github.com/lrstanley/go-ytdlp/blob/6224433347a07584fa888319430ccdd412a94e7b/post_processing.gen.go#L11-L35

Still have a bunch of things I'm testing/figuring out what makes the most sense for -- e.g. how to handle multi-choice options/enums, flags like `--version` and `--update`, etc.

I am also considering making an HTTP api wrapper/web UI (could auto-generate all of the validation/inputs/endpoints/etc) for invoking yt-dlp. I was originally planning on making a one-off for my specific usecase (currently have a webhook server call a bash script, which invokes yt-dlp), however, I think I could make it generic enough that anyone else could use it.

**Comment by alex-torok** [comment]
Also somewhat interested in this to provide help information for a yt-dlp GUI wrapper. The GUI shows what flag is used under the hood and I'd like to be able to show the help text value from yt-dlp itself when the `--some_flag` text is hovered.

> I have never seen any program provide such an output. Do you know of any?

A tool that I'm aware of that does this is the build system bazel. It has [hundreds to thousands of options](https://bazel.build/reference/command-line-reference?hl=en) and offers a [protobuf format output of its options](https://github.com/bazelbuild/bazel/blob/3ebd6adfdc836f456e0bcabac399f2e985c2933f/src/main/protobuf/bazel_flags.proto#L25-L62) (via the undocumented `bazel help flags-as-proto`). This allows downstream tooling to get structured information about all of the options that can be passed into bazel.

To steal additional inspiration from bazel, it also offers a hidden `bazel help everything-as-html` that outputs the docs in html format for the command line reference page that I linked above. 

The code for creating structured output for the yt-dlp options might also be useful for generating prettier documentation for yt-dlp flags.

@lrstanley - Did you continue to pursue this / would you be able to put up a PR with your implementation? (It seems like your `go-ytdlp` repo is recently active and somewhat popular - nice job!) I was going to open a PR that met my needs of extracting flag help descriptions, but my idea for an implementation here definitely wouldn't have had all of the information that it seems like you're trying to get for autogenerated library wrappers.

**Comment by lrstanley** [comment]
> [@lrstanley](https://github.com/lrstanley) - Did you continue to pursue this / would you be able to put up a PR with your implementation? (It seems like your `go-ytdlp` repo is recently active and somewhat popular - nice job!) I was going to open a PR that met my needs of extracting flag help descriptions, but my idea for an implementation here definitely wouldn't have had all of the information that it seems like you're trying to get for autogenerated library wrappers.

Yes. Each go-ytdlp version is pinned to a specific yt-dlp version, and I have a patch file [that I maintain here](https://github.com/lrstanley/go-ytdlp/tree/master/cmd/patch-ytdlp), which patches yt-dlp, generates the json output, then does the rest of the code-generation for my library based on it. The patch file has only needed to be updated a few times, mostly because lines around the patch were changed, not the actual logic itself.

There are some gotchas, in that things like deprecated or not-recommended flags aren't actually documented in code in the vast majority of cases (only in the readme), so I do have [some logic here](https://github.com/lrstanley/go-ytdlp/tree/master/cmd/codegen), which keeps track of those, but none of that logic is really required, it's just to make the generated library cleaner and more "exact" (not promoting the use of legacy flags and similar).

You can see an example of what that patch exports here: https://cdn.liam.sh/share/2025/09/export-2025.09.26.json

I can submit a PR with the patch easily, although I haven't heard from maintainers if this is something they want to support.

**EDIT**: also, I didn't even think about the idea of generating documentation for yt-dlp itself using the exported information, but that might actually help in making the code actually align with the README better, if the code is the source of truth for the bulk of the README, then having isolation sections for examples/more thorough information where warranted. Also makes maintainers' lives easier.

**Comment by bashonly** [comment]
> also, I didn't even think about the idea of generating documentation for yt-dlp itself using the exported information, but that might actually help in making the code actually align with the README better

The [USAGE AND OPTIONS](https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#usage-and-options) section of the README is [already](https://github.com/yt-dlp/yt-dlp/blob/master/.github/workflows/release.yml#L122) [generated](https://github.com/yt-dlp/yt-dlp/blob/master/Makefile#L106) [automatically](https://github.com/yt-dlp/yt-dlp/blob/master/devscripts/make_readme.py)
